### PR TITLE
Optimise Localities - Use the difference between old and new suburbs

### PR DESCRIPTION
**Pull request by MDavidson17**
Fixes: #371

### Change Description:
The process for updating suburbs and localities has been improved.

The new process to update localities:

1. Update the buildings that overlap localities where the geometries have changed
- This fixes buildings that overlap the areas where suburbs have expanded into.

The new function (buildings.building_outlines_update_changed_and_deleted_suburb()):
- Finds the ids and geometry differences where suburbs have changed shape
- Intersect the change with the current table (buildings_reference.suburb_locality) producing a subquery containing the 'old_id', 'new_id" and difference shape
- Updates building outlines to have the new_id where the building has the 'old_id' and overlaps the difference shape (by the largest proportion)

2. Insert new localities into buildings_reference.suburb_locality and update the overlapping buildings

The new function (buildings.building_outlines_update_added_suburb()):
- Adds the new suburbs to buildings_reference.suburb_locality (to get the new suburb_locality_ids)
- Finds the old suburb locality ids that overlap the new ids.
- Updates building outlines to have the new_od where the building has an old_id and it overlaps the new shape (by the largest proportion)

3. Deletes localities from buildings_reference.suburb_locality where the external id is no longer in admin_bdys.nz_locality

4. Modifies/updates existing localities where the attributes or the geometries have changed

### Notes for Testing:
- Test data has been added and the plugin tests updated.
- The process has not been tested on tstassgeo01 yet as it needs to be updated.

#### Source Code Documentation Tasks:
- [ ] README updated (where applicable)
- [ ] CHANGELOG (Unreleased section) updated
- [ ] Docstrings / comments included to help explain code

#### User Documentation Tasks:
- [ ] Confluence user guide updated (where applicable)

#### Testing Tasks:
- [ ] Added tests that fail without this change
- [ ] All tests are passing in development environment
- [ ] Reviewers assigned
- [ ] Linked to main issue for ZenHub board

### CODE REVIEW

File: buildings/gui/reference_data.py [hunk 1]
```diff
@@ -15,9 +15,7 @@
 from buildings.utilities import database as db
 
 __location__ = os.path.realpath(os.path.join(os.getcwd(), os.path.dirname(__file__)))
-FORM_CLASS, _ = uic.loadUiType(
-    os.path.join(os.path.dirname(__file__), "reference_data.ui")
-)
+FORM_CLASS, _ = uic.loadUiType(os.path.join(os.path.dirname(__file__), "reference_data.ui"))
```
Comment (dwsilk) [inline buildings/gui/reference_data.py]: Can we remove the formatting changes from this PR?

We need to apply Black consistently across the whole codebase but should do that in a separate PR.

File: db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    building_outline_updates
+    AS (UPDATE 
+        buildings.building_outlines
+        SET suburb_locality_id = ic.new_id
+    FROM 
+    intersected_change ic, buildings_reference.suburb_locality bsl
+    WHERE 
+        buildings.building_outlines.suburb_locality_id = bsl.suburb_locality_id
+        AND bsl.external_suburb_locality_id = ic.old_id
+        AND (ST_Within(ST_Expand(buildings.building_outlines.shape, 100), ic.shape)
```
Comment (Douglas-K) [inline db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql]: We previously used ST_Expand 1000m.  Consider changing back to this or ensure all buildings are within 100m of LOCALITY/ISLAND/SUBURB/PARK_RESERVE

File: db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        OR (ST_Overlaps(ST_Expand(buildings.building_outlines.shape, 100), ic.shape)
```
Comment (Douglas-K) [inline db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql]: We previously used ST_Expand 1000m.  Consider changing back to this or ensure all buildings are within 100m of LOCALITY/ISLAND/SUBURB/PARK_RESERVE

File: db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    update_building_outlines AS (
+        UPDATE 
+          buildings.building_outlines b
+        SET 
+          suburb_locality_id = ao.new_suburb_loc_id
+        FROM 
+          added_overlap ao
+        WHERE 
+          b.suburb_locality_id = ao.old_suburb_loc_id
+          AND (ST_Within(ST_Expand(b.shape,100), ao.shape)
```
Comment (Douglas-K) [inline db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql]: We previously used ST_Expand 1000m.  Consider changing back to this or ensure all buildings are within 100m of LOCALITY/ISLAND/SUBURB/PARK_RESERVE

File: db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          OR (ST_Overlaps(ST_Expand(b.shape,100), ao.shape)
```
Comment (Douglas-K) [inline db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql]: We previously used ST_Expand 1000m.  Consider changing back to this or ensure all buildings are within 100m of LOCALITY/ISLAND/SUBURB/PARK_RESERVE

File: db/sql/deploy/buildings_reference/functions/suburb_locality_intersect.sql [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+CREATE OR REPLACE FUNCTION buildings_reference.suburb_locality_area_of_second_largest_intersect(
+    p_polygon_geometry geometry, p_excluded_id integer
+)
+RETURNS double precision AS
+$$
+
+    SELECT ST_Area(ST_Intersection(p_polygon_geometry, shape))
+    FROM buildings_reference.suburb_locality
+    WHERE shape && ST_Expand(p_polygon_geometry, 100)
```
Comment (Douglas-K) [inline db/sql/deploy/buildings_reference/functions/suburb_locality_intersect.sql]: We previously used ST_Expand 1000m.  Consider changing back to this or ensure all buildings are within 100m of LOCALITY/ISLAND/SUBURB/PARK_RESERVE

File: db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql [hunk 7]
```diff
@@ -0,0 +1,141 @@
+-- Deploy nz-buildings:buildings/functions/update_suburb_locality_changed_deleted_buildings to pg
+
+BEGIN;
+
+-- building_outlines_update_changed_and_deleted_suburb (replace suburb values with the intersection result)
```
Comment (Douglas-K) [inline db/sql/deploy/buildings/functions/buildings_update_suburb_locality.sql]: Maybe add a comment to clarify that deleted localities aren't processed separately, but rather replaced by added and changed localities.
